File: Algorithms/BubbleSort.vb
Module Module1

    Sub Main()
        Dim swapped As Boolean
        Dim unsorted() As Integer
        Dim l, temp As Integer

        'Driver Code
        unsorted = {9, 94, 934, 3, 12, 7, 1, 88, 65, 4}
        l = unsorted.Length


        For i = 0 To l
            swapped = False

            ' -2 because in vb the for loop is lower bound to upper bound inclusive unlike python
            For j = 0 To (l - i - 2)

                If unsorted(j) > unsorted(j + 1) Then
                    temp = unsorted(j)
                    unsorted(j) = unsorted(j + 1)
                    unsorted(j + 1) = temp

                    swapped = True
                End If

            Next

            If swapped = False Then
                ' this is like break, if the array is sorted, the loop is ended to make 
                'code efficient
                i = l
            End If
        Next

        For i = 0 To l - 1
            Console.Write(Str(unsorted(i)) & " ")

        Next

        Console.ReadKey()

    End Sub

End Module

File: Algorithms/BubbleSort.py
# bubble sort function
def bubble_sort(my_list):
    length = len(my_list)
    
    # goes through all array elements
    for i in range(length):
        # initilizes the swapped to false 
        swapped = False 
        
        # -1 is done because array starts from 0 and we have used the length of list which is always 1 unit greater 
        # - i is done to make to code efficient as the last element of the array will be always sorted after one loop..
        # .. so we dont access it again
        # remember the for loop in python is lower bound inclusive, upper bound exclusive
        for j in range(length - i - 1):
            
            # if the number in current index is greater than the number in the index after that..
            # ..swap them so that greater number goes forward(right)
            if my_list[j] > my_list[j + 1]:
                my_list[j], my_list[j + 1] = my_list[j + 1], my_list[j]
                
                # if swap happens, this will become true
                swapped = True
                
        # if the swap did not happen. swapped will remain false hence the list/array is sorted
        # which means we dont need to loop through the array again ( this makes the algorithm efficient)
        if swapped == False:
            break
        
        
# Driver Code
my_list = [9,8,74,4,5,24,3,9,10,0,3]

# main program          
bubble_sort(my_list)
print("The Sorted array is: ", my_list)


"""
In Python lists are transfered to a function or copied to a variable as reference
so you don't need create another list and append the values into it
"""        